Lib/pathlib/_os.py

```diff
@@ -394,6 +242,102 @@ def ensure_different_files(source, target):
     raise err
 
 
+def copy_file(source, target, follow_symlinks=True, dirs_exist_ok=False,
+              preserve_metadata=False):
+    """
+    Recursively copy the given source ReadablePath to the given target WritablePath.
+    """
+    info = source.info
+    if not follow_symlinks and info.is_symlink():
+        target.symlink_to(source.readlink(), info.is_dir())
+        if preserve_metadata:
+            target._write_info(info, follow_symlinks=False)
+    elif info.is_dir():
+        children = source.iterdir()
+        target.mkdir(exist_ok=dirs_exist_ok)
+        for src in children:
+            dst = target.joinpath(src.name)
+            copy_file(src, dst, follow_symlinks, dirs_exist_ok, preserve_metadata)
+        if preserve_metadata:
+            target._write_info(info)
+    else:
+        if _copyfile2:
+            # Use fast OS routine for local file copying where available.
+            try:
+                source_p = os.fspath(source)
+                target_p = os.fspath(target)
+            except TypeError:
+                pass
+            else:
+                _copyfile2(source_p, target_p)
+                return
```

Comment:
In a previous PR, @AA-Turner [raised](https://github.com/python/cpython/pull/126263#discussion_r1874615110):

> Do we permit types inheriting from `pathlib._abc.PathBase` but not also `PurePath` (or: not implementing `__fspath__`)? This seems a little strange at first look.

That's right - the `ReadablePath.copy()` method supports an arbitrary `WritablePath` target. Neither source nor target are required to be `os.PathLike` - they could be paths in a zip file or on S3 etc.